text/3680-use.md

```diff
@@ -0,0 +1,422 @@
+- Feature Name: `use`
+- Start Date: 2024-07-20
+- RFC PR: [rust-lang/rfcs#3680](https://github.com/rust-lang/rfcs/pull/3680)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Provide a feature to simplify performing lightweight clones (such as of
+`Arc`/`Rc`), particularly cloning them into closures or async blocks, while
+still keeping such cloning visible and explicit.
```

Comment:
To be sure I am following, @joshtriplett or others, can you confirm that the key technical proposals in this RFC are as follows:

* Add a new trait `Use` intended to signal clones that meet 3 key criteria:
    * cheap (typically O(1), no more than a few cache lines);
    * transparent (the result of the clone is equivalent to the original after the clone is complete; there is nothing you can do after cloning you can't do before); and
    * infallible (i.e., the clone doesn't panic under typical conditions or assert weird invariants).
* Add new syntax `x.use`, which will
    * require that the type of `x` implements `Use`;
    * potentially invoke `clone`, but potentially not if compiler can avoid it:
        * If the place `x` is dead after the `use`, then this can be a move.
        * If a move is not allowed, the bytes will be memcopied the type implements `Copy`, otherwise `clone` will be called (**I'm not sure if this is included in the RFC; imo it should be.**).
        * Further, the compiler may opt to use a reference to `x` instead if it can see that this is safe.
            * This could be observable to users if a type embeds an `UnsafeCell` and implements `Use`. This is ok.
* Add new syntax `use || *body*` and `async use { *body* }`:
    * Values referenced in a "use" closure (and blocks) will either be moved (if this is the last use) or "used" otherwise (same as above).

Key points I want to double check:

* Use closures (and blocks) don't require *all* upvars to implement `Use`, only that that are still live after the closure. Upvars dead after closure creation will simply be moved and could be of any type.
* When `use`'ing a value of type `T`, if `T: Copy`, then this is always a memcpy (`clone` is not called). This is determined at monomorphization time.

Reply:
> Key points I want to double check:

> Use closures (and blocks) don't require all upvars to implement Use, only that that are still live after the closure. Upvars dead after closure creation will simply be moved and could be of any type.

That isn't the current definition in the RFC, but it's one worth considering.

Upside: This would handle non-Use values easily, such as arrays.
Downside: This would move things that the user might prefer to have remembered to wrap in an indirection (Box, Arc, ...) to reduce how much needs to get moved. (They probably couldn't just have been borrowed because otherwise the closure/block wouldn't have needed `use` at all, but the user might prefer to have remembered to wrap them.)

On balance, it's probably reasonable to change to this definition.

> When use'ing a value of type T, if T: Copy, then this is always a memcpy (clone is not called). This is determined at monomorphization time.

Not explicitly, but when `T: Copy`, clone should always be a memcpy. Is there some advantage to short-circuiting this rather than calling `Clone::clone` and expecting that to turn into a memcpy?

Reply:
I added the possibility of directly using `Copy` to the alternatives section for consideration. Would help to know more of what the advantage would be.

Reply:
Added the possibility of closures/blocks moving things that are dead.